server/src/main/java/org/elasticsearch/action/admin/cluster/reroute/TransportClusterRerouteAction.java

```diff
@@ -69,18 +87,82 @@ protected ClusterRerouteResponse newResponse() {
     @Override
     protected void masterOperation(final ClusterRerouteRequest request, final ClusterState state,
                                    final ActionListener<ClusterRerouteResponse> listener) {
-        ActionListener<ClusterRerouteResponse> logWrapper = ActionListener.wrap(
-            response -> {
-                if (request.dryRun() == false) {
-                    response.getExplanations().getYesDecisionMessages().forEach(logger::info);
+        // Gather all stale primary allocation commands into a map indexed by the index name they correspond to
+        // so we can check if the nodes they correspond to actually have any data for the shard
+        Map<String, List<AbstractAllocateAllocationCommand>> stalePrimaryAllocations = null;
+        for (AllocationCommand command : request.getCommands().commands()) {
+            if (command instanceof AllocateStalePrimaryAllocationCommand) {
+                if (stalePrimaryAllocations == null) {
+                    stalePrimaryAllocations = new HashMap<>();
```

Comment:
I think we can afford to make this `HashMap` eagerly and avoid the noise in the loop.